### PR TITLE
Remove `env_aliases` references from Meltano

### LINKED ISSUE #3048: Remove all `env_aliases` references from Meltano
Will be closed by #5927

Migrated from GitLab: https://gitlab.com/meltano/meltano/-/issues/3120

Originally created by @aaronsteers on 2021-12-14 18:52:17

---

Our latest `discovery.yml` file contains ~~36 refs to `env:` and~~ 73 references to `env_aliases:`. I propose we remove these from `discovery.yml` and instead direct users to use the explicit environment variable logic: `<plugin_name>_<setting_name>`.

Optionally, we can preserve this functionality in case users want to add their own mappings within their `meltano.yml` files, but to my understanding, this is fully redundant with just setting the config value to be equal to the value (e.g. `value: $MY_ENV_VAL` is basically equivalent to adding the `env: MY_ENV_VAL` or `env_aliases: [MY_ENV_VAL]`).

As I understand the evolution of this featureset, the custom-named env variable aliases existing prior to the newer `<plugin_name>_<setting_name>` parsing convention. Since adding that newer convention, I think we have seen this convention go mainstream for most use cases, and we run into problems with those legacy variable names being confusing user experience if the ingestion is not expected, and/or if the order of resolution is surprising to users. See [#2984](https://gitlab.com/meltano/meltano/-/issues/2984) for more info on this.

## Update (2022-01-24)

Following from the notes in [!2507](https://gitlab.com/meltano/meltano/-/merge_requests/2507), we think that `env:` is serving a different function, and so this deprecation should only affect `env_aliases`. Further discussion re: `env:` can happen in [!2507](https://gitlab.com/meltano/meltano/-/merge_requests/2507) and then we'll feed the findings/outcomes back into this issue.

## Best reason not to build

This is a breaking change for users who are using the older (custom) environment variable names for their projects. 

That said, it should be a simple fix to _either_ add the alias back to the `meltano.yml` file (as the `value:` or mapped in `env:`/`env_aliases:`) or just use the newer `<plugin_name>_<setting_name>` variable names.

### LINKED ISSUE #3120: Task Orchestration != Task Execution
Migrated from GitLab: https://gitlab.com/meltano/meltano/-/issues/3193

Originally created by @kgpayne on 2022-01-31 15:03:58

---

**Note:** this issue is a written record of thoughts and experiences to serve as background to specific proposals. It need not be read unless i) it is linked to and relevant to a specific proposal you are evaluating, or ii) you are especially curious about the specifics of data task orchestration and execution and have time to spend reading.

**tldr;** today we expect that to run data tasks at all in production, we need to deploy an 'orchestrator'. This need not be the case, and is even contrary to the intent and design of the data worlds most popular orchestrator - Airflow. We need more and better execution options, not more orchestrators.

## Historical Context 

I use Airflow as the base example for this issue as it is widely understood and also served as the inspiration for later orchestrators. In the 'Current State of the Art' section below I endeavour to cite similarities and differences from other popular orchestrators.

Airflow, developed internally at AirBnb, was [announced and Open Sourced in June of 2015](https://medium.com/airbnb-engineering/airflow-a-workflow-management-platform-46318b977fd8) and has since become a core part of many organisations data infrastructure. From the very beginning Airflow was built to "author, schedule, and monitor data pipelines", but not _execute_ the tasks in those pipelines. Back in 2015 AirBnb highlighted the **Extensibility** of the Airflow tool, through the use of Hooks, Operators and Executors:

> Hooks are defined as external systems abstraction and share a homogenous interface. Hooks use a centralized vault that abstracts host/port/login/password information and exposes methods to interact with these system.
>
> Operators leverage hooks to generate a certain type of task that become nodes in workflows when instantiated. All operators derive from BaseOperator and inherit a rich set of attributes and methods. 
> 
> Executors implement an interface that allow Airflow components (CLI, scheduler, web server) to run jobs jobs remotely. Airflow currently ships with a SequentialExecutor (for testing purposes), a threaded LocalExecutor, and a CeleryExecutor that leverages Celery, an excellent asynchronous task queue based on distributed message passing. We are also planning on sharing a YarnExecutor in the near future. 

Core to Airflows model then is an explicit design decision to "run jobs remotely" - to separate orchestration from execution. 9 months later, AirBnb shared details about how Airflow is used within their wider Data Infrastructure:

airbnb_data_infrastructure

> Scheduling of jobs happens through Airflow, a platform to programmatically author, schedule, and monitor data pipelines that can run jobs on Hive, Presto, Spark, MySQL, and more — note that we logically share Airflow across clusters, but physically the Airflow jobs run on the appropriate sets of clusters, machines, and workers. We built Airflow and open sourced it, as well. The Spark cluster is another processing tool favoured heavily by engineers and data scientists working on machine learning, and is useful for stream processing.

The team who conceived of and built the first iterations of Airflow emphasise over and over that "we logically share Airflow across clusters, but physically the Airflow jobs run on the appropriate sets of clusters, machines, and workers."

Fast forward a few years and it becomes clear this advice was not heeded. With convenient facilities to run arbitrary code within the Airflow execution environment (using the bulit-in `BashOperator` and `PythonOperator` operators, or by subclassing the `BaseOperator` directly), 1000's of companies not running AirBnb-scale workloads on Hive/Presto/Spark implemented their tasks directly in Airflow. Whilst convenient to begin with, this soon became problematic. Anyone working on an Airflow project created before 2019 is likely to have come across dependency tree and resource isolation issues that result from 'doing work' in Airflows execution environment:

- **Dependency tree issues**: All 3 of the offending operators execute within the same environment as Airflow itself. This means that any task dependencies (e.g. Pandas, AWS SDK, other 3rd Party SaaS clients) must not conflict with each other _or Airflow_. I personally remember the AWS case being particularly painful, with many SDK's using various versions of `botocore` to interact with common AWS integration points such as S3. Meltano, with our creation and management of virtualenvs, goes some way in alleviating this pain point, however we don't directly address resource isolation.

- **Resource isolation issues**: Another consequence of running arbitrary code and tasks within the same environment is the "noisy neighbours" effect. Airflows parallel execution strategy (when using the threaded `LocalExecutor` and the `CeleryExecutor`) is [optimistic](https://en.wikipedia.org/wiki/Optimistic_concurrency_control); Executors will execute tasks up to the configured `parallelism` (for `LocalExecutor`) or `worker_concurrency` (for `CeleryExecutor`) _without any additional control or consideration for resource consumption_. This core execution strategy is 'good enough' when job execution is offloaded onto "appropriate sets of clusters, machines, and workers"; all the workers are doing is polling an external job runner (such as Spark) to retrieve status and log information. However optimistic execution fails miserably when two Network-intensive (e.g. Extract-Load) or CPU-intensive (e.g. Transform) jobs are run concurrently directly on the Airflow workers. Jobs begin to hang or fail at random due to "transient" resource issues, affecting overall stability of users pipelines.

In summary, managing dependency and resource constraints of arbitrary data tasks in Airflow is not easy _by design_, as these use cases (however common) fall outside of Airflows intended use as a "job scheduler" _not_ a "job runner".

## Current 'State of the Art'

So how do people manage today? After a [concerted campaign](https://medium.com/bluecore-engineering/were-all-using-airflow-wrong-and-how-to-fix-it-a56f14cb0753) in the Airflow community to extricate tools from the Airflow environment, we have arrived at the `KubernetesPodOperator` (and equivalents for other managed Docker execution platforms such as the `EcsOperator`) as a Best Practice. This approach directs users to wrap tasks in separate Docker images and to launch them onto Docker-compatible clusters like Kubernetes or ECS using the provided operators. This solves the two problems stated above; dependency-isolation is managed using separate Docker images for each task (or task type), and Docker-compatible clusters have native constructs for specifying the resources required by those launched Dockerized tasks (CPU, memory, volume-mounts, secrets etc). Huzzah!

This shift to 'Dockerizing' tasks for execution is so successful that it is supported out-of-the-box by all of the 'second wave' orchestrators:

<details>
<summary><h3>Dagster</h3></summary>

[**Architecture**](https://docs.dagster.io/deployment/overview#architecture)

dagster-deployment

[**Supported Executors**](https://docs.dagster.io/deployment/executors)

Example executors include:

- `in_process_executor`: Execution plan executes serially within the run worker itself.
- `multiprocess_executor`: Each step executes within its own spawned process. Has configurable level of parallelism.
- `dask_executor`: Executes each step within a dask task.
- `celery_executor`: Executes each step within a celery task.
- `docker_executor`: Executes each step within a Docker container.
- `k8s_job_executor`: Executes each step within an ephemeral kubernetes pod.
- `celery_k8s_job_executor`: Executes each step within a ephemeral kubernetes pod, using celery as a control plane for prioritization and queuing.
- `celery_docker_executor`: Executes each step within a Docker container, using celery as a control plane for prioritization and queueing.

</details>

<details>
<summary><h3>Prefect</h3></summary>

[**Architecture**](https://docs.prefect.io/orchestration/#architecture-overview)

prefect_architecture_overview

[**Supported Agents**](https://docs.prefect.io/orchestration/agents/overview.html#agent-types)

Prefect supports several different agent types for deploying on different platforms.

- `Local`: The Local Agent executes flow runs as local processes.
- `Docker`: The Docker Agent executes flow runs in docker containers.
- `Kubernetes`: The Kubernetes Agent executes flow runs as Kubernetes Jobs.
- `GCP Vertex`: The Vertex Agent executes flow runs as Vertex Custom Jobs.
- `AWS ECS`: The ECS Agent executes flow runs as AWS ECS tasks (on either ECS or Fargate).

</details>

## Limitations of the Current State

Whilst new Dockerized execution patterns alleviate the issues of dependency tree and resource isolation issues of Airflows' past, they are not without their own challenges nor do they yet meet the aspirations of "jobs run on the appropriate sets of clusters, machines, and workers".

- Cluster managers expose _centralised_ resources, a pattern that is adequate for deploying web services but that is often not fit for the distributed nature of the data world. Using a single cluster for all the tasks in a DAG assumes all of the external service dependencies (e.g. APIs, databases, filestores etc.) of each Task in that DAG are met. For even a simple ELT pipeline, this assumes source and target are both accessible on the same network as the cluster. In a world of microservices, multi-VPC and multi-Account architecture patterns, this is a fragile assumption, and when false it forces teams to build and maintain complex network bridges to provide centralised access to those disparate services. Whilst it is possible to run two different DAGs (or their equivalent in the newer orchestrators) on two separate clusters, it is not readily possible to run two separate Tasks in the same DAG on separate clusters, nor is it attractive to run multiple clusters (with all the overhead that entails).

- Docker and available cluster management tooling (such as Kubernetes) can be a high price to pay for the benefits derived. It imposes a steep learning curve and administrative burden on data teams that is only partially mitigated by paid managed offerings.

- Taking a direct dependency on a cluster manager for the execution of DAGs compromises the local development experience (even as containerisation makes it easier to test individual Tasks). To execute a DAG made of `KubernetesPodOperator` tasks locally you must have a Kuberentes cluster deployed and available locally, with resources (such as volumes, secrets etc.) closely matching the configuration in production. This is not a simple task to set up or maintain.

- Whilst it is natural to build Spark apps in the Spark framework, for remote execution on a Spark cluster, expressing other kinds of Tasks in standalone Docker containers negates the benefits offered by Task definition decorators and base classes included with the major Orchestrators. No general-purpose Task definition SDK has achieved broad adoption, nor is there yet a standalone Task 'app store' for reuse of common Tasks. Defining and containerising Tasks is left as an exercise for the end-user. More on this in [Task Orchestration != Task Implementation]().

## Consequences for Meltano

By wrapping an Orchestrator as a proxy for providing Execution (i.e. `meltano invoke airflow` for Airflow to call `meltano elt ...`), we inherit the limitations of that Orchestrator. However, if we acknowledge that Orchestration != Execution, we can choose to reverse the relationship by providing our own Execution paradigm for Orchestrators (current and future) to integrate with, providing a Task Execution Framework (rather than a Task Orchestration Framework) performing a function equivalent to Spark in the big-data ecosystem. This idea is proposed in [#3139](https://gitlab.com/meltano/meltano/-/issues/3139).

### LINKED ISSUE #3429: Deprecate env_aliases and add second setting for MELTANO_DISABLE_TRACKING
Migrated from GitLab: https://gitlab.com/meltano/meltano/-/issues/3513

Originally created by @aaronsteers on 2022-05-19 21:05:32

---

The following discussion from [!2656](https://gitlab.com/meltano/meltano/-/merge_requests/2656) should be addressed:

- [ ] @kgpayne started a [discussion](https://gitlab.com/meltano/meltano/-/merge_requests/2656#note_953627644):  (+1 comment)

    > @tayloramurphy its worth calling out that, as `env_aliases` is going away, this means the `MELTANO_DISABLE_TRACKING` env var is going too (as its a negated alias of `send_anonymous_usage_stats`). Possibly not the only visible change here, but worth calling out specifically for comment 🙂 I caught up briefly with @aaronsteers and we talked about adding a separate setting called `disable_tracking`, or renaming the existing `send_anonymous_usage_stats` to `disable_tracking` as possibilities.

### PR DESCRIPTION
**Pull request by MeltyBot**
Migrated from GitLab: https://gitlab.com/meltano/meltano/-/merge_requests/2656

Originally created by @kgpayne on 2022-05-18 11:21:40

---

## Merge Request Checklist

- [x] Link to an issue
- [x] Include the proposed fix or feature
- [x] Include and update tests for the modified code
- [x] Include a documentation change
- [ ] Add a [CHANGELOG.md](https://gitlab.com/meltano/meltano/blob/master/README.md#Changelog) entry in the Unreleased section for any user-facing changes
- [x] Mention or assign a maintainer

### CODE REVIEW

File: src/meltano/core/plugin/project_plugin.py [hunk 1]
```diff
@@ -254,7 +254,7 @@ def env_prefixes(self, for_writing=False) -> list[str]:
         return uniques_in(prefixes)
 
     @property
-    def extra_config(self) -> list[str, Any]:
+    def extra_config(self) -> dict[str, Any]:
```
Comment (aaronsteers) [inline src/meltano/core/plugin/project_plugin.py:257]: For my info, is this a bugfix on the type declaration or did something else change here?

Comment (aaronsteers) [reply src/meltano/core/plugin/project_plugin.py:257]: @kgpayne - Was this ready for final review or still in progress?

There are some comments on the GitLab side that I'm not sure if/how they were resolved.

Namely:

1. Did you end up reverting the removal of the `env_aliases` feature and do both versions of the tracking ID still work?
2. This is missing a changelog entry, which would probably answer my question `1` above.
3. Did you get all the references found/replaced on other related MRs/PRs? (Linking to those would be helpful if there are related merges that are needed along with this one.)

If those three are addressed, I think this could be ready to go.

On the docs side:

- [ ] A mapping of environment variables to switch from for existing projects is needed for #3434 .

You could add that as a comment and TODO on #3434 since the upgrade guide page doesn't exist yet. Or you could add here and we'll resolve conflicts when we get to that point.

**Review by aaronsteers — CHANGES_REQUESTED** [review]

**Review by aaronsteers — APPROVED** [review]
Hi, @kgpayne - If you are able to clear the pytest CI actions, please go ahead and merge this if you are able. ✅ The migration docs can be added in #5975.

That said, if it is not ready to merge by US daytime hours, do you mind adding some comments about what next steps should be here? I may have some time during my time Sunday and would be happy to help move this forward.

Thanks!

File: docs/src/_contribute/prerequisites.md [hunk 2]
```diff
@@ -55,15 +55,13 @@ This means that you're ready to start Meltano CLI development. For API and UI de
 
 <div class="notification is-warning">
   <p><strong>Metrics (anonymous usage data) tracking</strong></p>
-  <p>As you contribute to Meltano, you may want to disable <a href="/reference/settings#send-anonymous-usage-stats">metrics tracking</a> globally rather than by project. You can do this by setting the environment variable `MELTANO_DISABLE_TRACKING` to `True`:</p>
+  <p>As you contribute to Meltano, you may want to disable <a href="/reference/settings#send-anonymous-usage-stats">metrics tracking</a> globally rather than by project. You can do this by setting the environment variable `MELTANO_SEND_ANONYMOUS_USAGE_STATS` to `False`:</p>
 <pre>
 # Add to `~/.bashrc`, `~/.zshrc`, etc, depending on the shell you use:
-export MELTANO_DISABLE_TRACKING=True
```
Comment (tayloramurphy) [inline docs/src/_contribute/prerequisites.md:61]: Confirming this is going to be added back in https://github.com/meltano/meltano/issues/3429

**Review by tayloramurphy — APPROVED** [review]
Minor question, docs look good to my eye.

File: docs/src/_contribute/prerequisites.md (hunk 2 above)
Comment (kgpayne) [reply docs/src/_contribute/prerequisites.md:61]: The understanding I got from the various discussions is that a `disable_tracking` setting would only be added for maintaining backwards compatibility for existing users, and so should not be a recommended way to configure the `send_anonymous_usage_stats` setting going forward. Adding it back would defeat that purpose _unless_ we intend to remove `send_anonymous_usage_stats` in favour of `disable_tracking` as the single setting in a future release. I am not sure if that has been decided, but either way I think we should only document one of the two options to make deprecation of the other as smooth as possible for new users 🙂

Comment (tayloramurphy) [reply docs/src/_contribute/prerequisites.md:61]: @kgpayne good point. I think we should add it to #5975 and #3434 for sure though 😄 thanks for clarifying!

**Comment by tayloramurphy** [comment]
Note we have https://github.com/meltano/hub/issues/547 to remove these from the Hub references as well.

**Comment by aaronsteers** [comment]
I'm re-locking to resolve the merge conflict and then will merge this.
